power_domain: Internal SoC PM state change

mmahadevan108 opened the pull request:
System managed device power management only invokes PM_DEVICE_ACTION_SUSPEND and PM_DEVICE_ACTION_RESUME actions on the device drivers. 
This changes allows the user to specify a property per power state called `zephyr,pm-turn-on-off-action`. This will make the system managed power management framework to invoke PM_DEVICE_ACTION_TURN_OFF and PM_DEVICE_ACTION_TURN_ON actions on the device drivers.
This allows device drivers to take additional power saving steps for certain low power states.

gmarull reviewed (dismissed):
these actions are meant to notify devices about being turned off or being turned on, please, check https://docs.zephyrproject.org/latest/services/pm/power_domain.html

mmahadevan108 commented:
This change is to allow devices and SoC's that are not in a power domain to receive TURN_ON/OFF actions. Its a property per power state that is in the users control. 
If there is a power domain driver then this property can be left as false.
Do we need the TURN_ON/OFF actions to always go through a power-domain driver?

JordanYates commented:
> Do we need the TURN_ON/OFF actions to always go through a power-domain driver?

`TURN_ON` and `TURN_OFF` are defined in terms of power (voltage) being physically applied or removed from the device in question, not as a lower power sleep state. Device drivers cannot properly implement these actions is they don't have a common meaning across use cases.

ceolin commented:
> > Do we need the TURN_ON/OFF actions to always go through a power-domain driver?
> 
> `TURN_ON` and `TURN_OFF` are defined in terms of power (voltage) being physically applied or removed from the device in question, not as a lower power sleep state. Device drivers cannot properly implement these actions is they don't have a common meaning across use cases.

Yep that is correct. But it is possible that depending on power state that the SoC sleeps, internal devices my have power removed.

ceolin commented:
Maybe the nomenclature is ambiguous ? If this action is *only* used in cases where a power state cuts of device power, I don't see a problem.

mmahadevan108 commented:
We have SoC low power states where the IP blocks lose state and need to be reconfigured on exit from low power modes. `SUSPEND` and `RESUME` does not work in this and the device driver needs to receive a different action. I am using TURN_ON as that seemed an appropriate one.
Do you suggest we add a different action?

JordanYates commented:
> We have SoC low power states where the IP blocks lose state and need to be reconfigured on exit from low power modes.

I don't use system PM, is there a link somewhere between the IP blocks that do lose state and the power state with `zephyr,pm-turn-on-off-action`?
Or will this result in the action being run on all devices?

> Do you suggest we add a different action?

This is tricky, because adding an extra state means you have to fit it into the existing [state transitions](https://docs.zephyrproject.org/latest/services/pm/device.html#device-power-management-states).

I think the current `TURN_ON` and `TURN_OFF` states are doing double duty to signify both the voltage applied (GPIO configuration needed) and device state (Reconfiguration needed) actions. Which mostly makes sense, because for external devices these two tend to be linked. If the device is powered it can hold state, if it is not it cannot.

Are these IP blocks losing state because they are being powered down, or for some other reason (i.e. clocks disabled)?

mmahadevan108 commented:
> > We have SoC low power states where the IP blocks lose state and need to be reconfigured on exit from low power modes.
> 
> I don't use system PM, is there a link somewhere between the IP blocks that do lose state and the power state with `zephyr,pm-turn-on-off-action`? Or will this result in the action being run on all devices?
> 
> > Do you suggest we add a different action?
> 
> This is tricky, because adding an extra state means you have to fit it into the existing [state transitions](https://docs.zephyrproject.org/latest/services/pm/device.html#device-power-management-states).
> 
> I think the current `TURN_ON` and `TURN_OFF` states are doing double duty to signify both the voltage applied (GPIO configuration needed) and device state (Reconfiguration needed) actions. Which mostly makes sense, because for external devices these two tend to be linked. If the device is powered it can hold state, if it is not it cannot.
> 
> Are these IP blocks losing state because they are being powered down, or for some other reason (i.e. clocks disabled)?

They are being power down by the SoC.

JordanYates commented:
> They are being power down by the SoC.

Then I don't see a problem with `TURN_ON` and `TURN_OFF` being used.

> This action will be run on all devices. If for some reason the device is already TURNED_OFF, then the system will ignore and move on to the next device. I could add logic to skip calling the TURN_OFF action on devices that are registered with a power-domain driver.

This is a problem. Your suggested solution is not sufficient because that would also run the action on external devices that are not on a power domain. IMO there needs to be a setup where the actions are run ONLY on the IP blocks that are being powered down, not relying on the PM subsystem to discard actions.

ceolin commented:
> > They are being power down by the SoC.
> 
> Then I think `TURN_ON` and `TURN_OFF` can be used with no problems.

Yep !

> 
> > This action will be run on all devices. If for some reason the device is already TURNED_OFF, then the system will ignore and move on to the next device. I could add logic to skip calling the TURN_OFF action on devices that are registered with a power-domain driver.
> 
> This is a problem. Your suggested solution is not sufficient because that would also run the action on external devices that are not on a power domain. IMO there needs to be a setup where the actions are run ONLY on the IP blocks that are being powered down, not relying on the PM subsystem to discard actions.

Agree, this action should run only in devices that are losing power. @mmahadevan108 you can use the property added in this commit 02a14d75fc7. It tells which power states cause power loss in a device. You may leverage it and trigger on/off in devices using it.

mmahadevan108 commented:
> > > They are being power down by the SoC.
> > 
> > 
> > Then I think `TURN_ON` and `TURN_OFF` can be used with no problems.
> 
> Yep !
> 
> > > This action will be run on all devices. If for some reason the device is already TURNED_OFF, then the system will ignore and move on to the next device. I could add logic to skip calling the TURN_OFF action on devices that are registered with a power-domain driver.
> > 
> > 
> > This is a problem. Your suggested solution is not sufficient because that would also run the action on external devices that are not on a power domain. IMO there needs to be a setup where the actions are run ONLY on the IP blocks that are being powered down, not relying on the PM subsystem to discard actions.
> 
> Agree, this action should run only in devices that are losing power. @mmahadevan108 you can use the property added in this commit [02a14d7](https://github.com/zephyrproject-rtos/zephyr/commit/02a14d75fc7fc0c1894f7832ce20fc48edf5ac66). It tells which power states cause power loss in a device. You may leverage it and trigger on/off in devices using it.

@ceolin, the `zephyr,disabling-power-states` will disable Device PM for a particular power state and I do not want that to happen. We do want device PM to be called, however with a different action.

mmahadevan108 commented:
> > They are being power down by the SoC.
> 
> Then I think `TURN_ON` and `TURN_OFF` can be used with no problems.
> 

Great, this is what this PR is proposing.
> > This action will be run on all devices. If for some reason the device is already TURNED_OFF, then the system will ignore and move on to the next device. I could add logic to skip calling the TURN_OFF action on devices that are registered with a power-domain driver.
> 
> This is a problem. Your suggested solution is not sufficient because that would also run the action on external devices that are not on a power domain. IMO there needs to be a setup where the actions are run ONLY on the IP blocks that are being powered down, not relying on the PM subsystem to discard actions.

In the change proposed, it is assumed that for certain power states all devices except the one's marked for wakeup are powered down.

mmahadevan108 commented:
@JordanYates @ceolin @gmarull please let me know your thoughts/suggestions on this PR.

JordanYates commented:
> In this proposal, it is assumed that for certain power states all devices except the one's marked for wakeup are powered down.

That is an incorrect assumption, you can't make any statements about the power state of devices that are external to the SoC.

mmahadevan108 committed:
drivers: power_domain: Add a driver to trigger TURN_ON/TURN_OFF actions

This driver triggers the TURN_ON and TURN_OFF actions for certain
power states. These power states are specified via device tree.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

JordanYates reviewed (changes requested):
Running `PM_DEVICE_ACTION_TURN_OFF` on every device in the application (including those external to the SoC) when the SoC transitions to a sleep state is not an acceptable solution.

mmahadevan108 commented:
Apologies for the delay in uodating this.
@JordanYates , @gmarull . I have changed the approach to invoke the TURN_ON and TURN_OFF actions from a power domain driver.

JordanYates reviewed (changes requested):
The driver should not be named for what it is doing, since every power domain driver turns devices on and off. It is performing actions in response to internal power state changes, the naming should reflect that.

mmahadevan108 commented:
@JordanYates, I have updated the naming of the file names

mmahadevan108 commented:
@ceolin , can you help revisit this PR. Thank you.

mmahadevan108 commented:
@JordanYates @gmarull @ceolin can you kindly help review.

JordanYates commented on drivers/power_domain/CMakeLists.txt:
```diff
@@ -7,3 +7,5 @@ zephyr_library_sources_ifdef(CONFIG_POWER_DOMAIN_GPIO power_domain_gpio.c)
 zephyr_library_sources_ifdef(CONFIG_POWER_DOMAIN_GPIO_MONITOR power_domain_gpio_monitor.c)
 zephyr_library_sources_ifdef(CONFIG_POWER_DOMAIN_INTEL_ADSP power_domain_intel_adsp.c)
 zephyr_library_sources_ifdef(CONFIG_POWER_DOMAIN_NXP_SCU power_domain_nxp_scu.c)
+zephyr_library_sources_ifdef(CONFIG_POWER_DOMAIN_STATE_CHANGE_DEVICEONOFF
```
Single line please

mmahadevan108 replied:
Changed.

JordanYates commented on drivers/power_domain/Kconfig:
```diff
@@ -90,4 +90,13 @@ config POWER_DOMAIN_NXP_SCU_INIT_PRIORITY
 
 endif #POWER_DOMAIN_NXP_SCU
 
+config POWER_DOMAIN_STATE_CHANGE_DEVICEONOFF
```
```suggestion
config POWER_DOMAIN_SOC_PM_STATE
```

mmahadevan108 replied:
Updated

JordanYates commented on drivers/power_domain/Kconfig:
```diff
@@ -90,4 +90,13 @@ config POWER_DOMAIN_NXP_SCU_INIT_PRIORITY
 endif #POWER_DOMAIN_NXP_SCU
 
+config POWER_DOMAIN_STATE_CHANGE_DEVICEONOFF
+	bool "Turn on/off devices during a PM state change"
```
```suggestion
	bool "SoC PM state power domain"
```

mmahadevan108 replied:
Changed.

JordanYates commented on dts/bindings/power-domain/power-domain-state-change-deviceonoff.yaml:
```diff
@@ -0,0 +1,20 @@
+  in and out a specified Power State. Power States that trigger this
+  action is specified via the onoff-power-states property.
+
+compatible: "power-domain-state-change-deviceonoff"
```
`deviceonoff` should be dropped from all naming, since that is just the definition of what a power domain does.

mmahadevan108 replied:
Updated.

JordanYates commented on drivers/power_domain/power_domain_state_change_device_onoff.c:
```diff
@@ -0,0 +1,116 @@
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
```
You've gone to the effort of defining a LOG module, it would be nice to add a `LOG_INF("%s: suspending", dev->name);` or something here and in the `PM_DEVICE_ACTION_RESUME` handling

mmahadevan108 replied:
Thanks for the suggestion. I have added a few.

JordanYates reviewed (changes requested):
Sorry for the continuing new requests, I just realized the driver isn't used anywhere.
There needs to be at least one test or sample in-tree so that CI makes sure the driver compiles.

mmahadevan108 commented:
@JordanYates, I can pull some parts of this PR into this to show usage. Would that work?
https://github.com/zephyrproject-rtos/zephyr/pull/84945

JordanYates commented:
> @JordanYates, I can pull some parts of this PR into this to show usage. Would that work? #84945

https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/pm/power_states_api demonstrates that it is quite simple to test the PM state API in CI. I would prefer a copy of that test which adds in your power domain driver, and then does some basic validation that the devices are put into the correct states as the SoC PM states transition.

That provides the advantage of also testing that the implementation works, not just that the code compiles.

ceolin reviewed (changes requested):
Minor revisions made by me. I believe this aligns with our desired direction, which is to map all devices to their respective power rails. It is logical to have a separate power rail for devices within the same SoC power domain.

ceolin commented:
> > @JordanYates, I can pull some parts of this PR into this to show usage. Would that work? #84945
> 
> https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/pm/power_states_api demonstrates that it is quite simple to test the PM state API in CI. I would prefer a copy of that test which adds in your power domain driver, and then does some basic validation that the devices are put into the correct states as the SoC PM states transition.
> 
> That provides the advantage of also testing that the implementation works, not just that the code compiles.

Agree !

mmahadevan108 committed:
tests: pm: Test the SoC State Change Power Domain driver

The SoC State Change Power Domain driver issues TURN_ON/
TURN_OFF actions to all devices registered with it for
certain power states that can be specified via device tree.
This test exercises the functionality of this driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

JordanYates commented:
> I think the issue highlighted with the original approach was it was calling TURN_ON and TURN_OFF actions on every device. This can be avoided by adding a pm_device_on_power_domain check and skip devices that are managed by a Power Domain Driver. The TURN_ON/OFF would still get called on the PD driver as there is no way I know to identify it.
Do you think this approach would work?

No, that would still run actions on external devices that are not on a power domain (permanently powered).

JordanYates commented:
> One thing with this new approach of using a Power Domain Driver is that every node in the SoC dtsi file would need to be updated to add the power-domains property.

That sounds like a reasonable requirement, there needs to be some link between devices and the power states they work in, where else would you put it?

mmahadevan108 commented on drivers/power_domain/power_domain_soc_state_change.c:
```diff
@@ -0,0 +1,118 @@
+LOG_MODULE_REGISTER(power_domain_deviceonoff, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+/* Indicates the end of the onoff_power_states array for count purposes */
+#define POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT 0xFF
```
Updated the macro name.

mmahadevan108 commented on drivers/power_domain/power_domain_soc_state_change.c:
```diff
@@ -0,0 +1,118 @@
+	};
+
+#define POWER_DOMAIN_DEVICE(id)								\
+	POWER_DOMAIN_DEVICE_ONOFF_STATES(id, DT_DRV_INST(id))				\
```
Updated

mmahadevan108 commented:
> [`main`/tests/subsys/pm/power_states_api](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/subsys/pm/power_states_api) demonstrates that it is quite simple to test the PM state API in CI. I would prefer a copy of that test which adds in your power domain driver, and then does some basic validation that the devices are put into the correct states as the SoC PM states transition.
> 
> That provides the advantage of also testing that the implementation works, not just that the code compiles.
Test added.

mmahadevan108 commented:
@gmarull , could you kindly revisit this PR.

mmahadevan108 commented:
> > I think the issue highlighted with the original approach was it was calling TURN_ON and TURN_OFF actions on every device. This can be avoided by adding a pm_device_on_power_domain check and skip devices that are managed by a Power Domain Driver. The TURN_ON/OFF would still get called on the PD driver as there is no way I know to identify it.
> > Do you think this approach would work?
> 
> No, that would still run actions on external devices that are not on a power domain (permanently powered).

Thank you. For my reference and understanding, can you point me to some of these devices

JordanYates commented:
> Thank you. For my reference and understanding, can you point me to some of these devices

Pick practically any sensor driver, and find an instance of it in a board file without `power-domains = <>;`.
As a random example, https://github.com/zephyrproject-rtos/zephyr/blob/759e31d0dc96441a0ac7847d71b674644070d167/boards/nordic/thingy53/thingy53_nrf5340_common.dtsi#L221

ceolin commented:
@mmahadevan108 thanks for putting it together. I tried the test and the failure does not seem to be a test problem. I noticed one additional tick in the timeout that is making the kernel call twice the suspend function instead of one. I am checking why this is happening.

mmahadevan108 commented:
> @mmahadevan108 thanks for putting it together. I tried the test and the failure does not seem to be a test problem. I noticed one additional tick in the timeout that is making the kernel call twice the suspend function instead of one. I am checking why this is happening.

Thanks @ceolin, I have added a change to address that issue.

ceolin commented on tests/subsys/pm/power_domain_soc_state_change/app.overlay:
```diff
@@ -0,0 +1,46 @@
+};
+
+&cpu0 {
+	cpu-power-states = <&state2 &state1 &state0>;
```
the states order is wrong, you should go from lower to max residency time.

mmahadevan108 replied:
Fixed

ceolin commented on tests/subsys/pm/power_domain_soc_state_change/src/main.c:
```diff
@@ -0,0 +1,130 @@
+ */
+const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
+{
+	const struct pm_state_info *cpu_states;
```
You should change the policy to use `ticks` to return the appropriate  state or having a better control over the states you are returning. Using `testing_domain_on_notitication` here is fragile since the policy can be called more than once during the sleep due external interruption or another scheduled event.

mmahadevan108 replied:
Fixed

ceolin commented on tests/subsys/pm/power_domain_soc_state_change/src/main.c:
```diff
@@ -0,0 +1,130 @@
+	k_timer_init(&idle_timer, idle_timer_expiry_function, NULL);
+
+	/* Sleep to transition to first STATE: STANDBY */
+	k_sleep(Z_TIMEOUT_TICKS(1));
```
Can you use the values declared in DTS for residency time ? It is more clear

mmahadevan108 replied:
Fixed

mmahadevan108 commented:
@ceolin @JordanYates @gmarull , thank you for your reviews. I think I have addressed your review comments. Can you please take a look one more time.

ceolin commented:
@JordanYates can you take another look please ?

JordanYates commented on tests/subsys/pm/power_domain_soc_state_change/src/main.c:
```diff
@@ -0,0 +1,93 @@
+	/* Sleep to transition to STATE: STANDBY */
+	k_usleep(state->min_residency_us + state->exit_latency_us);
+
+	zassert_equal(testing_domain_on_notitication, 1);
```
Why does transitioning from one state to another invoke both the `PM_DEVICE_ACTION_TURN_ON` and `PM_DEVICE_ACTION_TURN_OFF` notifications?
